### PR TITLE
Initial Contour Controller Implementation

### PR DESCRIPTION
**Pull request by danehans**
- Moves `main()` to cmd/operator pkg. `cmd/operator/start.go `: Exposes a flag to set the Contour image used by the operator.
- Singularizes the controller pkg.
- `controller/contour/controller.go`: Initial controller implementation that watches the contour CRD and a) finalizes the resource and b) creates the contour namespace if it does not exist.
- `util/retryableerror/retryableerror.go`: Utility function for retrying reconciliation errors.
- `util/slice/slice.go`: Utility function to check if a given slice of strings contains the provided string.

Signed-off-by: Daneyon Hansen <daneyonhansen@gmail.com>

### CODE REVIEW

File: cmd/operator/start.go [hunk 1]
```diff
@@ -76,9 +83,9 @@ func main() {
 	}
 	// +kubebuilder:scaffold:builder
 
-	setupLog.Info("starting manager")
+	setupLog.Info("starting contour-operator")
 	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
-		setupLog.Error(err, "problem running manager")
+		setupLog.Error(err, "problem running contour-operator")
```
Comment (jpeach) [inline cmd/operator/start.go:88]: nit: "failed to start contour-operator" would be a marginally better error

File: controller/contour/controller.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+// Reconciler reconciles a Contour object.
+type Reconciler struct {
+	Config
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=operator.projectcontour.io,resources=contours,verbs=get;list;watch;create;update;patch;delete
```
Comment (jpeach) [inline controller/contour/controller.go:57]: Do we need `create;update;patch;delete` access on the main resource?

File: controller/contour/namespace.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+package contour
+
+import (
+	"context"
+	"fmt"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
```
Comment (jpeach) [inline controller/contour/namespace.go:26]: Group imports by stdlib, then everything else.

File: util/retryableerror/retryableerror_test.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		},
+	}
+	for _, test := range tests {
+		err := NewMaybeRetryableAggregate(test.errors)
+		if retryable, gotRetryable := err.(Error); gotRetryable != test.expectRetryable {
+			t.Errorf("%q: expected retryable %T, got %T: %v", test.name, test.expectRetryable, gotRetryable, err)
+		} else if gotRetryable && retryable.After() != test.expectAfter {
+			t.Errorf("%q: expected after %v, got %v: %v", test.name, test.expectAfter, retryable.After(), err)
+		}
+		if _, gotAggregate := err.(utilerrors.Aggregate); gotAggregate != test.expectAggregate {
```
Comment (jpeach) [inline util/retryableerror/retryableerror_test.go:75]: Nit: `testify/assert` is pretty good

File: util/retryableerror/retryableerror_test.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+			errors: []error{
+				New(errors.New("baz"), time.Second*15),
+				New(errors.New("quux"), time.Minute),
+				nil,
+			},
+			expectRetryable: true,
+			expectAfter:     time.Second * 15,
+		},
+	}
+	for _, test := range tests {
```
Comment (jpeach) [inline util/retryableerror/retryableerror_test.go:68]: You can use subtests here:
```Go
cases := map[string]struct{ ... } { ... }
for name, case := range cases {
   t.Run(name, func(t *testing.T) {
      err := NewMaybeRetryableAggregate(case.errors)
      ...
   }
}
```

File: controller/contour/controller.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+			// stale queue entries (or something edge triggering from a related
+			// resource that got deleted async).
+			r.Log.Info("contour not found; reconciliation will be skipped", "request", req)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get contour %q: %v", req, err)
+	}
+
+	// The contour is safe to process, so ensure current state matches desired state.
+	desired := true
```
Comment (jpeach) [inline controller/contour/controller.go:80]: I guess it's ok for now, but 

1. passing a bool is harder to read and understand than an enum
2. teardown probably needs to be done in the reverse order

File: controller/contour/namespace.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+// ensureNamespace ensures the namespace for the provided name exists.
+func (r *Reconciler) ensureNamespace(desired bool, name string) (*corev1.Namespace, error) {
+	ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: name}}
+	if err := r.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, ns); err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, fmt.Errorf("failed to get namespace %s: %v", ns.Name, err)
```
Comment (jpeach) [inline controller/contour/namespace.go:34]: This isn't an error if we are tearing down the namespace, though we wold want to update status in that case (maybe just add a TODO here?)

**Review by jpeach — CHANGES_REQUESTED** [review]

File: controller/contour/controller.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+	if err := r.ensureContour(desired, contour); err != nil {
```
[the remark this answers is not in the document]
Comment (danehans) [reply controller/contour/controller.go:82]: `ensureContour` is the func that will encapsulate all the lower-level funcs required to ensure dependent resources exist. A mirrored workflow will be created for removing contour and dependent resources. I've been working through the changes and will submit a new commit shortly. The pattern I'm trying to establish is similar to other operators and upstream k8s controllers. I'm open to feedback, so let me know your thoughts after I reviewing my upcoming commit.

File: controller/contour/controller.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/projectcontour/contour-operator/util/slice"
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	operatorv1alpha1 "github.com/projectcontour/contour-operator/api/v1alpha1"
```
[the remark this answers is not in the document]
Comment (danehans) [reply controller/contour/controller.go]: I'm for stdlib, local, vendor. I'll clean up the imports with my upcoming commit.

[the remark this answers is not in the document]
Comment (jpeach) [reply controller/contour/controller.go]: Cool, I believe that `goimports` has a `-local` flag that may help with this; maybe we can add that to the lint config :)

File: controller/contour/controller.go (hunk 6 above)
Comment (danehans) [reply controller/contour/controller.go]: Yeah, I refactored based on https://github.com/projectcontour/contour-operator/pull/10#discussion_r481492143.

File: controller/contour/namespace.go (hunk 7 above)
Comment (danehans) [reply controller/contour/namespace.go:34]: ack, fixed.

File: util/retryableerror/retryableerror_test.go (hunk 4 above)
Comment (danehans) [reply util/retryableerror/retryableerror_test.go:75]: I created issue https://github.com/projectcontour/contour-operator/issues/13 to give this decision the proper treatment.

File: Dockerfile [hunk 10]
```diff
@@ -10,18 +10,18 @@ COPY go.sum go.sum
...
 # Copy the go source
-COPY main.go main.go
+COPY cmd/operator/start.go start.go
 COPY api/ api/
-COPY controllers/ controllers/
+COPY controller/ controller/
 
 # Build
-RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o start start.go
```
[the remark this answers is not in the document]
Comment (danehans) [reply Dockerfile:18]: ack

File: cmd/operator/start.go (hunk 1 above)
Comment (danehans) [reply cmd/operator/start.go:88]: ack

File: controller/contour/controller.go (hunk 2 above)
Comment (danehans) [reply controller/contour/controller.go:47]: I'll change to `get;list;watch;update`. We'll most likely have future use cases that will require the additional verbs, but the they can be updated as needed.

**Comment by danehans** [comment]
@jpeach I believe commit `e83f7b1` addresses all your feedback except https://github.com/projectcontour/contour-operator/pull/10#discussion_r480759628. I've never implemented subtests like you mention. Do you have a Contour example to share or would you mind implementing after this merges so I can learn?

**Review by jpeach — APPROVED** [review]
LGTM

**Pull request merged**